src/sp/it/pl/plugin/LibraryWatcher.kt

```diff
@@ -0,0 +1,152 @@
+package sp.it.pl.plugin
+
+import mu.KLogging
+import org.reactfx.Subscription
+import sp.it.pl.audio.SimpleItem
+import sp.it.pl.audio.tagging.MetadataReader
+import sp.it.pl.main.APP
+import sp.it.pl.main.showAppProgress
+import sp.it.pl.service.notif.Notifier
+import sp.it.pl.util.action.IsAction
+import sp.it.pl.util.async.executor.EventReducer
+import sp.it.pl.util.async.runFX
+import sp.it.pl.util.async.runNew
+import sp.it.pl.util.collections.materialize
+import sp.it.pl.util.conf.EditMode.NONE
+import sp.it.pl.util.conf.IsConfig
+import sp.it.pl.util.conf.c
+import sp.it.pl.util.conf.cList
+import sp.it.pl.util.conf.cv
+import sp.it.pl.util.conf.only
+import sp.it.pl.util.conf.readOnlyUnless
+import sp.it.pl.util.file.AudioFileFormat
+import sp.it.pl.util.file.FileMonitor
+import sp.it.pl.util.file.FileMonitor.monitorDirectory
+import sp.it.pl.util.file.Util
+import sp.it.pl.util.file.isChildOf
+import sp.it.pl.util.reactive.Subscribed
+import sp.it.pl.util.reactive.onItemAdded
+import sp.it.pl.util.reactive.onItemRemoved
+import sp.it.pl.util.reactive.sync
+import sp.it.pl.util.system.Os
+import sp.it.pl.util.text.plural
+import sp.it.pl.util.validation.Constraint.FileActor.DIRECTORY
+import java.io.File
+import java.nio.file.StandardWatchEventKinds.ENTRY_CREATE
+import java.nio.file.StandardWatchEventKinds.ENTRY_DELETE
+import kotlin.streams.toList
+
+class LibraryWatcher: PluginBase("Song Library", false) {
+
+    @IsConfig(name = "Location", info = "Locations to find audio.")
+    private val sourceDirs by cList<File>().only(DIRECTORY)
+    private val sourceDirsChangeHandler = Subscribed {
+        sourceDirs.forEach { handleLocationAdded(it) }
+        Subscription.multi(
+                sourceDirs.onItemAdded { handleLocationAdded(it) },
+                sourceDirs.onItemRemoved { handleLocationRemoved(it) }
+        )
+    }
+
+    @IsConfig(name = "Update on start", info = "Update library when this plugin starts")
+    val updateOnStart by cv(false)
+
+    @IsConfig(name = "Monitoring supported", info = "On some system, this file monitoring may be unsupported", editable = NONE)
+    val dirMonitoringSupported by c(Os.WINDOWS.isCurrent)
+
+    @IsConfig(name = "Monitor files", info = "Monitors files recursively, notify of changes and update library automatically")
+    val dirMonitoringEnabled by cv(false).readOnlyUnless(dirMonitoringSupported)
+
+    private val dirMonitors = HashMap<File, FileMonitor>()
+    private val dirMonitoring = when {
+        dirMonitoringSupported -> Subscribed {
+            Subscription.multi(
+                    dirMonitoringEnabled sync { sourceDirsChangeHandler.subscribe(it) },
+                    Subscription { sourceDirsChangeHandler.subscribe(false) }
+            )
+        }
+        else -> Subscribed { Subscription { } }
+    }
+    private val toBeAdded = HashSet<File>()
+    private val toBeRemoved = HashSet<File>()
```

Comment:
I recommend to, for all Sets and Maps only expose the Set/Map API by setting the variable type, e.g. `private val toBeRemoved: Set<File> = HashSet<File>()`

Reply:
I agree with you if the property is public. For privates its just boilerplate.

Reply:
But even on private stuff you could potentially change the type of Collection you use, and if everything works fine with a lower exposure of implementation detail it could ease switching.

Reply:
I think we can allow a bit of leniency on private stuff.
```
private val toBeAdded = HashSet<File>()
```
```
private val toBeAdded: MutableSet<File> = HashSet<File>()
```
Meh.
It does not buy anything. A lot less readable. I see no reason to encapsulate the encapsulated. Let the innards see each other.

Btw hiding HashSets and ArrayLists behind Sets and Lists is not always good idea. If anything, by the rule of covariant return type, one should return the most specific type, not the most generic. Encapsulation is going the opposite direction. The end result is something in between. Like, you would hardly suggest to use `MutableCollection<File>` here right? Yet isn't that the result of the logic you applied here? The fact that 99% people use List/Set in their API instead of Collection or ArrayList appears like the monkey-see-monkey-do kind of story.

Anyway, usually one does not expose collections as properties, so the exact type is broadened to List/Set by method signature. On the other hand with private collection properties it is fine to let compiler evaluate property to be of the exact collection type and even let the internal use this fact to their advantage if possible. Its simple, its easy, its readable. Widening property type explicitly seems bloated and overprotective to me more often than not.